reco_utils/evaluation/python_evaluation.py

```diff
@@ -23,663 +23,347 @@
 )
 
 
-def check_column_dtypes(f):
-    """
-    Checks columns of dataframe inputs.
+class Evaluation:
```

Comment:
hey @gramhagen we had a discussion about the metrics interfaces months ago. We decided to follow the sklearn/TF/Pytorch convention of using functions instead of classes for Python. In PySpark we will follow the PySpark convention of using classes. There are multiple reasons to follow this:
1) We want to follow a data-driven software design [based on evidence](https://github.com/Microsoft/Recommenders/wiki/Coding-Guidelines#evidence-based-software-design). In this case the evidence is that sklearn/TF/Pytorch uses this convention. A user that uses sklearn will find natural to use functions in our metrics.
2) There's no need for classes when using metrics, you just need to make an operation with 2 df. Abusing classes increases complexity and reduce readability.
3) As @loomlike noted, we discussed that when performing some metrics, we are repeat certain operations. The way for maintaining single responsibility and at the same time being efficient is be using cache (same as we do in the github metrics). This is something that we haven't done so far and would be nice to implement.
4) We also thought about changing the PySpark interfaces and make then more pythonic (which would be having functions instead of classes), however, we left it as classes. Here the same principle applies, a PySpark user will find more natural to use classes on everything (we are not saying that this is the recommended approach, in fact it is not, but PySpark code have a strong Java background so to follow their convention we code in this line)

Reply:
I think the general principle is to make the interface and design pattern (if there is 😂) mimic the existing packages we are using. In case there is no existing approaches, we will look for evidence to support any proposed idea. 

Regarding the replicated calculation of metrics, it makes sense to have some preliminary steps (e.g., data join) before certain metrics of the same type are calculated to eliminate redundant computation. 

Would love to have everyone's opinion.

Reply:
I understand, and agree with the concept of using existing tools / apis. While I don't feel that this implementation is an abuse of python classes as it actually simplifies the current usage I think @loomlike has the right idea. We can reduce redundant operations (i.e. greatly speed up evaluation) by exposing the methods for extracting the truth and prediction arrays. Honestly at that point there's little need for any of our wrapper methods as we can use sklearn methods directly.

I can refactor this towards that goal if that sounds reasonable to everyone.

Reply:
We still need to keep the wrappers, because sklearn does not have all the metrics we want (or that we will build in the future). Also, the one that sklearn has may be unsuitable (e.g. precision), because it was not designed with recommendation use cases in mind.

But I agree we should delegate `merge_rating_true_pred` and `merge_ranking_true_pred` to the user, for efficiency.

Reply:
> One way to improve the efficiency (without making them as a class) is to make evaluation functions to accept pre-calculated intermediate dataframes (outputs of merge_ranking_true_pred()) that can be shared across those functions. If those dataframes are provided, skip the re-calculation.

Not sure if you are proposing to have a preprocessing step and then call the interface that we have. Something like:

```
x_true, x_pred = preprocess(df_true, df_pred)
p = precision_at_k(x_true, x_pred)
```
is this what @loomlike  and @gramhagen  are saying?

Reply:
It is more like 
`merged_true_pred, df_hits = preprocess_ranking(df_true, df_pred)
p = precision_at_k(merge_true_pred, df_hits)`

Reply:
@miguelgfierro Rather, introducing additional arg for cached df so that users can use the evaluation functions as usual **or** pre-compute the ranked-df and provide it to ranking evals as follows:
```
def precision_at_k(..., ranking_true_pred=None):
    if ranking_true_pred is None:
        ranking_true_pred = merge_ranking_true_pred(...)
    
    _, df_hit, n_users = ranking_true_pred

    ....

>>> cached_ranking = merge_ranking_true_pred(...)
>>> p_k = precision_at_k(..., cached_ranking)
>>> n_k = ndcg_at_k(..., cached_ranking)
...
````

If we want to use this approach, can we rename `merge_ranking_true_pred` function to be more descriptive? I think naming after the goal of the function make more sense rather than how the function achieve that goal. We can describe about 'how' part at the docstring. 

Also adding "Examples:" to all ranking evaluation function docstrings will be great!

Reply:
Also, it probably makes sense for `cached_ranking` to be a class rather than a tuple, since it will be a black box for the users anyway.

Reply:
ok what I'm thinking has the same result but doesn't need to add another argument, it would be something like this:

```
@lru_cache()
def merge_ranking_true_pred(df_true, df_pred):
    ...

def precision_at_k(df_true, df_pred):
    ranking_true_pred = merge_ranking_true_pred(...)
    ...
    ...

>>> p_k = precision_at_k(df_true, df_pred )
>>> n_k = ndcg_at_k(df_true, df_pred)
```
Note that the only thing I added was the decorator, this is catching the computation the first time it is done. Second time merge_ranking_true is called, it gets the cached value. I've never done it with code, we we have this in the repo metrics: https://github.com/Microsoft/Recommenders/blob/master/scripts/repo_metrics/git_stats.py#L61 and I guess it should be the same.

I think @msalvaris has done something similar in the past, is this doable?

Reply:
Lru cache stores inputs as keys in a dict, so they need to be hashable. I believe this will be a problem for dataframes and mutable arrays. There may be other cache implementations that provide workarounds? We should time the whole process to figure out where bottlenecks really are. Could be caching isn't a huge help, if we can avoid recalling the merging.

Reply:
Lru cache stores inputs as keys in a dict, so they need to be hashable. I believe this will be a problem for dataframes and mutable arrays. There may be other cache implementations that provide workarounds? We should time the whole process to figure out where bottlenecks really are. Could be caching isn't a huge help, if we can avoid recalling the merging.

Reply:
Dataframes should be hashable, we have used lru_cache with dataframes before. But I agree that only add caching when you know you need it.

Reply:
In [14]: a = np.array([1,2,3])

In [15]: hash(a)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-57b555d30865> in <module>
----> 1 hash(a)

TypeError: unhashable type: 'numpy.ndarray'

In [16]: df = pd.DataFrame(data={'a': 1, 'b': 'ewiu'}, index=[0])

In [17]: hash(df)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-17-5e54663ef3fb> in <module>
----> 1 hash(df)

~/anaconda3/envs/reco_pyspark/lib/python3.6/site-packages/pandas/core/generic.py in __hash__(self)
   1814     def __hash__(self):
   1815         raise TypeError('{0!r} objects are mutable, thus they cannot be'
-> 1816                         ' hashed'.format(self.__class__.__name__))
   1817
   1818     def __iter__(self):

TypeError: 'DataFrame' objects are mutable, thus they cannot be hashed

Reply:
https://docs.python.org/3.1/glossary.html
"Objects which are instances of user-defined classes are hashable by default; they all compare unequal, and their hash value is their id()"
So, we could wrap the dataframes inside a class we define, but this would not be a good idea, because:
- if you change the values in the encapsulated dataframe, the hash does not change (I double checked)
- it becomes implicit and obfuscates the user.

Reply:
but have you tried to check if lru_cache works with that function? that would just accelerate things and the code would be the same

Reply:
https://docs.python.org/3/library/functools.html
"Since a dictionary is used to cache results, the positional and keyword arguments to the function must be hashable."

Reply:
Thanks for checking @anargyri. Evindently forgot what we did :D.  You can use hash_pandas_object and simply implement lru_cache for dfs

Reply:
https://stackoverflow.com/questions/31567401/get-the-same-hash-value-for-a-pandas-dataframe-each-time

Reply:
For this discussion, see this PR that speeds up the rating computation https://github.com/Microsoft/Recommenders/pull/707 and this issue to improve it even further https://github.com/Microsoft/Recommenders/issues/708